src/main/java/at/ac/tuwien/kr/alpha/solver/ChoiceManager.java

```diff
@@ -147,13 +189,63 @@ public void updateAssignments() {
 		}
 	}
 
-	public void nextDecisionLevel() {
+	public void guess(Choice choice) {
+		guesses++;
+		LOGGER.debug("Guess {} is {}@{}", guesses, choice, assignment.getDecisionLevel());
+
+		if (assignment.guess(choice.getAtom(), choice.getValue()) != null) {
+			throw oops("Picked choice is incompatible with current assignment");
+		}
+
+		if (debugWatcher != null) {
+			debugWatcher.runWatcher();
+		}
+
 		highestDecisionLevel++;
 		modifiedInDecisionLevel.put(highestDecisionLevel, new ArrayList<>());
+
+		choiceStack.push(choice);
+	}
+
+	public void backjump(int target) {
+		if (target < 0) {
+			throw oops("Backjumping to decision level less than 0");
+		}
+
+		backjumps++;
+		LOGGER.debug("Backjumping to decision level {}.", target);
+
+		// Remove everything above the target level, but keep the target level unchanged.
+		while (assignment.getDecisionLevel() > target) {
+			backtrackFast();
+		}
+	}
+
+	public void backtrackFast() {
```

Comment:
What is the difference between `backtrackFast`, `backtrackSlow`, and `backtrack`? When is the one used and when the other? Please put that in some explanatory comment.

Reply:
`backtrack` is private and contains what is common to `backtrackSlow` and `backtrackFast`: It sets counters and recomputes active choice points.

`backtrackSlow` will read and return the assignment that resulted in the choice. This is to enable further analysis. It is called in `DefaultSolver.backtrack` where different actions are taken depending on what happened since the assignment was made.

`backtrackFast` just backtracks the assignment without caring about what the particular assignment was.

I will add a similar explanation in the code.